### PR TITLE
Implement ETag support

### LINKED ISSUE #4594: Etag support
`If-None-Match` header takes precedence over `If-Modified-Since` and allows caching http clients to refresh the content only if content etag changes. Right now, only aiohttp server supports only `If-Modified-Since` header but not `If-None-Match`

💡 **Describe the solution you'd like**
If GET/HEAD request contains  `If-None-Match`, the server should check for the resource etag and, it the etag matches the value in `If-None-Match` header, return `Not Modified` 304. More details are in [RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.26)

❓ **Describe alternatives you've considered**
None

### PR DESCRIPTION
**Pull request by greshilov**
## What do these changes do?
This PR adds `etag` property to the response object and `if_match`, `if_none_match` properties to the request object. Also it implements `ETag` support in static routes and fixes few bugs found along the way.

Despite the low priority of static files handling for `aiohttp`, I believe `ETag` and related headers has a wide application beyond this purpose.

## Are there changes in behavior for the user?

Nope, no general changes in server behaviour.

## Related issue number

Fixes #4594

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

---
This change is [Reviewable](https://reviewable.io/reviews/aio-libs/aiohttp/5298)

### CODE REVIEW

File: aiohttp/helpers.py
Comment (asvetlov) [inline aiohttp/helpers.py]:
```suggestion
_ETAGC = "[{}]+".format(
    "".join(
```

Comment (greshilov) [reply aiohttp/helpers.py]: Thanks! Missed that.

File: aiohttp/web_request.py
Comment (serhiy-storchaka) [inline aiohttp/web_request.py:512]: There is some inconsistency here. If ETag is `"*"`, the value attribute will contain quotes. In all other cases they will be stripped.

File: aiohttp/web_response.py
Comment (serhiy-storchaka) [inline aiohttp/web_response.py]: This test is the cheapest. I would move it to the start.

File: aiohttp/web_response.py
Comment (serhiy-storchaka) [inline aiohttp/web_response.py:304]: Maybe check that the value is valid (does not contain non-printable characters or `"`)?

File: aiohttp/web_response.py
Comment (serhiy-storchaka) [inline aiohttp/web_response.py]: `{type(value)!r}` is the same as `type(value)`. Maybe you meant `{type(value).__name__!r}`?

File: aiohttp/helpers.py
Comment (asvetlov) [reply aiohttp/helpers.py]: I prefer the first suggestion: don't want to fight with regexps when non-ASCII symbol appears in a tested string somehow

**Comment by asvetlov** [comment]
@serhiy-storchaka thank you very much for the review!

@greshilov please address comments, they are very valuable.

File: aiohttp/web_request.py
Comment (greshilov) [reply aiohttp/web_request.py]: That's a lot better, thanks!

File: aiohttp/web_request.py
Comment (greshilov) [reply aiohttp/web_request.py:515]: Yeah, you're right.

To solve this problem I rejected the `LIST_QUOTED_ETAG_RE` approach, and switched to `split` + `re.fullmatch`.
Now only `"1"` would be matched from proposed garbage string. I've checked `nginx` behaviour and it's consistent with suggested solution. `nginx` matches etag values from header until first invalid is found. New approach is also 10x faster than old.

Main disadvantage is that memory consumption will probably increase though, but I don't think that **really long** `If-Match, If-None-Match` headers are so widespreaded.

File: aiohttp/helpers.py
Comment (greshilov) [reply aiohttp/helpers.py]: I've decided to use `QUOTED_ETAG_RE` instead of `LIST_QUOTED_ETAG_RE`. See detailed justification few conversations above.

File: aiohttp/web_request.py
Comment (greshilov) [reply aiohttp/web_request.py]: Good point, thanks!

File: aiohttp/web_request.py
Comment (greshilov) [reply aiohttp/web_request.py:512]: Not quite. `*` is a kind of special value, which is used **without** quotes. So if someone somehow will use `"*"` instead of `*`, it will be considered as regular etag.

Unfortunatly `ETag(value='*')` produced from `"*"` is indistinguishable from one produced from `*`.
So, for example:
```
"*", "tag", "other-tag" -> (ETag(value='*'), ETag(value='other-tag'))
* -> (ETag(value='*'), )
```
Maybe it's better to create separate type `ETAG_ANY` and change `Optional[Tuple[ETag, ...]]` to `Optional[Union[ETAG_ANY, Tuple[ETag, ...]]]`
```
"*", "tag", "other-tag" -> (ETag(value='*'), ETag(value='other-tag'))
* -> ETAG_ANY
```

@asvetlov what do you think?

File: aiohttp/web_response.py
Comment (greshilov) [reply aiohttp/web_response.py]: Yep!

File: aiohttp/web_response.py
Comment (greshilov) [reply aiohttp/web_response.py:304]: Sounds reasonable. Done!

File: aiohttp/web_response.py
Comment (greshilov) [reply aiohttp/web_response.py]: Yeah, that's exactly what I meant. I guess better solution will be remove `!r`, because library code mostly contains `'%r' % type(smth)`.

File: aiohttp/web_response.py
Comment (greshilov) [reply aiohttp/web_response.py]: Done.

File: aiohttp/web_response.py
Comment (greshilov) [reply aiohttp/web_response.py]: Fixed, thanks.

**Comment by greshilov** [comment]
@serhiy-storchaka, @asvetlov thanks for the review! I appreciate your help.

I marked conversations as resolved where I used proposed solution without any changes, sorry if I shouldn't have done it by myself.

File: aiohttp/helpers.py
Comment (serhiy-storchaka) [reply aiohttp/helpers.py]: What encoding is used for decoding headers? How bytes `\x80-\xff` are mapped?

BTW there are needed tests for non-ASCII ETags.

File: aiohttp/__init__.py
Comment (serhiy-storchaka) [inline aiohttp/__init__.py]: Why not simply `from .helpers import BasicAuth, ChainMapProxy, ETag`?

File: aiohttp/web_request.py
Comment (serhiy-storchaka) [reply aiohttp/web_request.py:512]: You are right.

File: aiohttp/web_request.py
Comment (serhiy-storchaka) [inline aiohttp/web_request.py]: What if ETag contains comma? `"1,2"`

Actually the approach with LIST_QUOTED_ETAG_RE was not bad. You need just add `|(.)` at the end of the regexp. That group will be matched only if there is a garbage, so you can check it and raise an exception.
```
LIST_QUOTED_ETAG_RE = re.compile(fr"{_QUOTED_ETAG}(?:\s*,\s*|$)|(.)")
```

Comment (greshilov) [reply aiohttp/web_request.py]: Yeah, completely forgot about this case while thinking about optimizations. Tests were updated to reflect this situation.

Thanks for the smart solution!

File: aiohttp/__init__.py
Comment (greshilov) [reply aiohttp/__init__.py]: I guess it's a kind of outer interface protection measures. Andrew can tell for sure.

Comment (asvetlov) [reply aiohttp/__init__.py]: No, we have no policy for such cases.
Usually, we use `as` only for importing under a different name if there is a conflict.

Please feel free to drop `as`.

Comment (asvetlov) [inline aiohttp/__init__.py]:
```suggestion
from .helpers import (
    BasicAuth,
    ChainMapProxy,
    ETag,
)
```

File: aiohttp/__init__.py
Comment (serhiy-storchaka) [inline aiohttp/__init__.py]:
```suggestion
from .helpers import BasicAuth, ChainMapProxy, ETag
```

Comment (greshilov) [reply aiohttp/__init__.py]: Fixed.

File: aiohttp/helpers.py
Comment (greshilov) [reply aiohttp/helpers.py]: Non-ASCII test cases successfully added.

**Comment by greshilov** [comment]
I also have to mention, that according to rfc7232 `If-Range` header may contain `ETag` value. Current PR ignores that fact for now, because implementing `ETag` support for the `If-Range` header is not backward compatible, and should be separated to a different PR.

**Comment by asvetlov** [comment]
> I also have to mention, that according to rfc7232 `If-Range` header may contain `ETag` value. Current PR ignores that fact for now, because implementing `ETag` support for the `If-Range` header is not backward compatible, and should be separated to a different PR.

Agree

**Review by asvetlov — APPROVED** [review]
LGTM.
Sorry for delay in the review

**Comment by asvetlov** [comment]
@greshilov please merge.

Please edit the commit message first: press "Squash and merge" -> edit the text to cleanup meaningless records about intermediate commits -> press "Confirm squash and merge".

**Comment by greshilov** [comment]
> @greshilov please merge.
> 
> Please edit the commit message first: press "Squash and merge" -> edit the text to cleanup meaningless records about intermediate commits -> press "Confirm squash and merge".

Unfortunately, github doesn't show me this button. It says only users with write access can do this. 

But I can squash commits manually before merge, if needed!

**Comment by greshilov** [comment]
@webknjaz, @Dreamsorcerer can you help me with squash/merging this? Andrew approved this PR but he has no spare time now.

I'm also planning to do a backport to 3.8 this week.

File: aiohttp/__init__.py
Comment (Dreamsorcerer) [inline aiohttp/__init__.py:41]: Why are the imports changed here?

I'm assuming the style in this file follows the no implicit imports approach from Mypy:
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport

Although a little odd that it's not enforced in the Mypy tests if that is the case.

Comment (greshilov) [reply aiohttp/__init__.py:41]: Those classes are already presented in the `__all__`  variable. 
I agree about inconsistency with `mypy` style flags.

This change was previously discussed, I'll tag you there.

File: aiohttp/__init__.py
Comment (greshilov) [reply aiohttp/__init__.py]: @Dreamsorcerer

File: aiohttp/__init__.py
Comment (webknjaz) [reply aiohttp/__init__.py:41]: This is a great example of why style/formatting changes should be separated from the functional ones: a tiny disagreement can block a whole lot of legitimate changes, plus this makes it harder to notice important behavior changes.

I'll just leave this here: https://mtlynch.io/code-review-love/.

File: aiohttp/web_fileresponse.py
Comment (webknjaz) [inline aiohttp/web_fileresponse.py:113]: Nitpick, this could be less nested:
```suggestion
        return any(etag.value == etag_value for etag in etags if not etag.is_weak)
```

File: aiohttp/web_fileresponse.py
Comment (webknjaz) [inline aiohttp/web_fileresponse.py:121]: It was agreed in other PRs to add specific error codes in brackets.

File: docs/spelling_wordlist.txt
Comment (webknjaz) [inline docs/spelling_wordlist.txt:32]: I'm almost sure that the spellchecker is case-insensitive.

File: tests/test_web_request.py
Comment (webknjaz) [inline tests/test_web_request.py:827]: FTR it's usually better readable (though lesser known) to use iterables in parameterize:
```suggestion
    ("header", "header_attr"),
```

File: tests/test_web_response.py
Comment (webknjaz) [inline tests/test_web_response.py:306]: The best practice is to always include `, match=r'...'` argument demonstrating the intended exception reason more precisely.

**Comment by webknjaz** [comment]
@greshilov I didn't do a thorough review but commented with a few generic advices in several places. None of them are critical and since Andrew has already approved the patch, I'm going to merge it now. Feel free to send improvements in follow-up PRs.

**Comment by webknjaz** [comment]
@greshilov waiting for the backport to 3.8 now.

**Comment by greshilov** [comment]
@webknjaz sure, thank you!

**Pull request merged**
